src/logplex_api_v3_drains.erl

```diff
@@ -0,0 +1,238 @@
+         to_json/2,
+         delete_resource/2
+        ]).
+
+-record(state, {
+          channel_id :: binary(),
+          drain_id   :: undefined | logplex_drain:id(),
+          drain      :: undefined | logplex_drain:drain(),
+          uri        :: undefined | #ex_uri{} %% optional json url parameter
+         }).
+
+
+
+%% @private
+init(_Transport, Req, Opts) ->
+    Route = proplists:get_value(route, Opts),
+    Req1 = logplex_api_v3:prepare(Route, Req),
+    {upgrade, protocol, cowboy_rest, Req1, Opts}.
+
+%% @private
+rest_init(Req, _Opts) ->
+    {ChannelId, Req1} = cowboy_req:binding(channel_id, Req),
+    case cowboy_req:binding(drain_id, Req1) of
+        {undefined, Req2} ->
+            State = #state{ channel_id = ChannelId },
+            {ok, Req2, State};
+        {DrainId, Req2} ->
+            State = #state{ channel_id = ChannelId,
+                            drain_id   = DrainId },
+            {ok, Req2, State}
+    end.
+
+%% @private
+service_available(Req, State) ->
+    logplex_api_v3:service_available(Req, State).
+
+%% @private
+allowed_methods(Req, #state{ drain_id = undefined } = State) ->
+    %% No drain id on path, hence this is a POST request to create a new drain.
+    {[<<"POST">>], Req, State};
+allowed_methods(Req, State) ->
+    %% We have a drain id, hence this is an update request to a drain resource.
+    {[<<"PUT">>, <<"DELETE">>], Req, State}.
+
+%% @private
+malformed_request(Req, State) ->
+    {ok, Body, Req1} = cowboy_req:body(Req),
+    try jsx:decode(Body, [return_maps]) of
+        Map ->
+            case validate_payload(Map) of
+                {ok, URI} ->
+                    NewState = State#state{ uri = URI },
+                    {false, Req1, NewState};
+                {error, _Reason} ->
+                    {true, Req1, State}
+            end
+    catch
+        error:badarg ->
+            %% invalid json
+            {true, Req1, State}
+    end.
+
+%% @private
+is_authorized(Req, State) ->
+    logplex_api_v3:is_authorized(Req, State).
+
+%% @private
+resource_exists(Req, #state{ channel_id = ChannelId,
+                             drain_id   = undefined } = State) ->
+    %% this is a POST
+    Timeout = logplex_app:config(default_redis_poll_ms, 2000),
+    case logplex_channel:poll(ChannelId, Timeout) of
+        Channel when is_record(Channel, channel) ->
+            %% we return false here because we're attempting to create a new resource
+            {false, Req, State};
+        {error, timeout} ->
+            %% channel was not found
+            {ok, Req1} = cowboy_req:reply(404, Req),
+            {halt, Req1, State}
+    end;
+resource_exists(Req, #state{ channel_id = ChannelId,
+                             drain_id   = DrainId } = State) ->
+    Timeout = logplex_app:config(default_redis_poll_ms, 2000),
+    case logplex_channel:poll(ChannelId, Timeout) of
+        Channel when is_record(Channel, channel) ->
+            case logplex_drain:poll(DrainId) of
```

Comment:
Same comment as before, except this time we could hang out waiting for worst case 4 seconds.